src/vdaf.rs

```diff
@@ -166,9 +166,10 @@ pub trait Vdaf: Clone + Debug {
 pub trait Client: Vdaf {
     /// Shards a measurement into a public share and a sequence of input shares, one for each
     /// Aggregator.
-    fn shard(
+    fn shard<const N: usize>(
```

Comment:
Rather than using a const generic, could we instead use an [associated constant](https://doc.rust-lang.org/reference/items/associated-items.html#associated-constants)?

The issue with const generics is that they leak into code that is generic over the trait. For example, check out the many instances of `const L: usize` in [Janus](https://github.com/divviup/janus/blob/main/aggregator/src/aggregator.rs). I think using an associated const would avoid this issue.

An associated const would also make it easier to give a meaningful descriptive name to the constant -- right now we have `L` on `Aggregator` and now `N` on `Client`; we could rename these to (e.g.) `VERIFY_KEY_SIZE` and `NONCE_SIZE`, to aid understandability.

Reply:
The main advantage of using an const generic instead of an associated constant is that the size of the buffer for the nonce can be specified by the API. (To my knowledge, you can't have `&[u8; Self::NONCE_SIZE]` be a type for trait method.) This avoids the edge case where the user supplies `shard()` with a `nonce: &[u8]` whose length is not `Self::NONCE_SIZE`. In fact, this is why we have `const L: usize` for the verification key.

I'd be cool with an associated constant, but I'd like to suggest that we be consistent.

Reply:
Namely, we could add:

```
pub trait Vdaf:
   const VERIFY_KEY_SIZE: usize;
   const NONCE_SIZE: usize;
```

Reply:
For one thing, nobody said we had to use `N` as the name of the parameter. We can just as easily make it `const NONCE_SIZE: usize` for clarity.

Further, I agree that the value of a const generic is that then the compiler can statically know what the size of `[u8; NONCE_SIZE]` is. Otherwise I think you end up with unnecessary allocations because you have to go through `Vec::new` to get a `&[u8; Self::NONCE_SIZE]`. But I'm happy to be proven wrong; I haven't tried it myself.

So I lean toward a const generic, given my current understanding of this.

Reply:
Here is a minimal example:
```rust
pub trait Vdaf {
    const NONCE_SIZE: usize;
    fn shard(nonce: &[u8; Self::NONCE_SIZE]);
}
```

Neither stable nor nightly allows you to use an associated type this way:

```
0:20:18:~ $ rustc hella.rs
error: generic parameters may not be used in const operations
 --> hella.rs:3:27
  |
3 |     fn shard(nonce: &[u8; Self::NONCE_SIZE]);
  |                           ^^^^^^^^^^^^^^^^ cannot perform const operation using `Self`
  |
  = note: type parameters may not be used in const expressions

error[E0601]: `main` function not found in crate `hella`
 --> hella.rs:4:2
  |
4 | }
  |  ^ consider adding a `main` function to `hella.rs`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0601`.
```

Reply:
So if we wanted to use an associated type, then we'd pass in a `&[u8]` instead of a `&[u8; NONCE_SIZE]`.

Reply:
That's unfortunate -- I hope they expand Rust to support this, for this reason:

> For one thing, nobody said we had to use N as the name of the parameter. We can just as easily make it const NONCE_SIZE: usize for clarity.

The problem here is that the name is redefined at every point the const generic is used -- so even if `libprio-rs` sensibly names the constant something like `NONCE_SIZE`, implementations have to either (a) switch to using a less-descriptive name like `N` or (b) repeat the longer `NONCE_SIZE` name at every point-of-use of the related trait. Neither is great, and perhaps even worse is that both permit application code to get the name _wrong_, which will start to matter more as we introduce more related constants.

Reply:
Yeah there's not much we can do about that. We can at least nudge users in the right direction.

It sounds like we all might agree on two possibilities:
(1) Rename `L`  to `VERIFY_KEY_SIZE` and add `NONCE_SIZE` as const generic (downside: Users can rename these)
(2) Remove `L` and add `VERIFY_KEY_SIZE` and `NONCE_SIZE` as associated constants (downside: length checks are delegated to implementations of `Vdaf`)

Reply:
I'm in favor of (1) I think.

Reply:
Agreed, I am in favor of (1) unless/until we can use associated constants in method signatures.

Reply:
For future reference, using a generic associated constant in array lengths is currently behind the `#![feature(generic_const_exprs)]` gate in nightly, to an extent.